deeplearning4j-cuda/src/main/java/org/deeplearning4j/nn/layers/recurrent/CudnnLSTMHelper.java

```diff
@@ -0,0 +1,648 @@
+            HostDeallocator(DataCache c) {
+                super(c);
+            }
+
+            @Override
+            public void deallocate() {
+                checkCuda(cudaFreeHost(this));
+                setNull();
+            }
+        }
+
+        DataCache() {}
+
+        DataCache(long size) {
+            position = 0;
+            limit = capacity = size;
+            int error = cudaMalloc(this, size);
+            if (error != cudaSuccess) {
+                log.warn("Cannot allocate " + size + " bytes of device memory (CUDA error = " + error
+                                + "), proceeding with host memory");
+                checkCuda(cudaMallocHost(this, size));
+                deallocator(new HostDeallocator(this));
+            } else {
+                deallocator(new Deallocator(this));
+            }
+        }
+
+        DataCache(DataCache c) {
+            super(c);
+        }
+    }
+
+    static class TensorArray extends PointerPointer<cudnnTensorStruct> {
+
+        static class Deallocator extends TensorArray implements Pointer.Deallocator {
+            Pointer owner;
+
+            Deallocator(TensorArray a, Pointer owner) {
+                this.address = a.address;
+                this.capacity = a.capacity;
+                this.owner = owner;
+            }
+
+            @Override
+            public void deallocate() {
+                for (int i = 0; i < capacity; i++) {
+                    cudnnTensorStruct t = this.get(cudnnTensorStruct.class, i);
+                    checkCudnn(cudnnDestroyTensorDescriptor(t));
+                }
+                owner.deallocate();
+                owner = null;
+                setNull();
+            }
+        }
+
+        TensorArray() {}
+
+        TensorArray(long size) {
+            PointerPointer p = new PointerPointer(size);
+            p.deallocate(false);
+            this.address = p.address();
+            this.limit = p.limit();
+            this.capacity = p.capacity();
+
+            cudnnTensorStruct t = new cudnnTensorStruct();
+            for (int i = 0; i < capacity; i++) {
+                checkCudnn(cudnnCreateTensorDescriptor(t));
+                this.put(i, t);
+            }
+            deallocator(new Deallocator(this, p));
+        }
+
+        TensorArray(TensorArray a) {
+            super(a);
+        }
+    }
+
+    // These constants might eventually become variable parameters...
+    static final int numLayers = 1;
+    static final float dropout = 0;
+    static final boolean bidirectional = false;
+    static final int RNNMode = CUDNN_LSTM;
+    static final int numLinearLayers = 8; // CUDNN_LSTM
+
+    static int dataType = Nd4j.dataType() == DataBuffer.Type.DOUBLE ? CUDNN_DATA_DOUBLE
+                        : Nd4j.dataType() == DataBuffer.Type.FLOAT ? CUDNN_DATA_FLOAT : CUDNN_DATA_HALF;
+    static int dataTypeSize = Nd4j.dataType() == DataBuffer.Type.DOUBLE ? 8
```

Comment:
I'm a little worried about these static fields... I can imagine in tests: for example, CudnnLSTMHelper is initialized in a gradient check (double precision) then used again later in a different unit test (with float precision)...